File: code/functionapp/Json.cs
using common;
using LanguageExt;
using LanguageExt.Common;
using System;
using System.IO;
using System.Text.Encodings.Web;
using System.Text.Json;
using System.Text.Json.Nodes;
using System.Threading;

using static LanguageExt.Prelude;

namespace functionapp;

public static class JsonModule
{
    public static Error ToError(this JsonException jsonException)
    {
        return GetJsonError(jsonException.Message);
    }

    public static Error GetJsonError(string errorMessage)
    {
        return Error.New(CommonErrorCode.InvalidJson, errorMessage);
    }
}

public static class JsonObjectModule
{
    public static uint GetUIntProperty(this JsonObject jsonObject, string propertyName)
    {
        return jsonObject.TryGetUIntProperty(propertyName)
                         .Run()
                         .ThrowIfFail();
    }

    public static Eff<uint> TryGetUIntProperty(this JsonObject jsonObject, string propertyName)
    {
        return jsonObject.TryGetJsonValueOfTProperty<uint>(propertyName);
    }

    public static NonEmptyString GetNonEmptyStringProperty(this JsonObject jsonObject, string propertyName)
    {
        return jsonObject.TryGetNonEmptyStringProperty(propertyName)
                         .Run()
                         .ThrowIfFail();
    }

    public static Eff<NonEmptyString> TryGetNonEmptyStringProperty(this JsonObject jsonObject, string propertyName)
    {
        return jsonObject.TryGetStringProperty(propertyName)
                         .Bind(value => string.IsNullOrWhiteSpace(value)
                                        ? FailEff<NonEmptyString>(JsonModule.GetJsonError($"Property '{propertyName}' is a null, empty, or whitespace string."))
                                        : SuccessEff(new NonEmptyString(value)));

    }

    public static string GetStringProperty(this JsonObject jsonObject, string propertyName)
    {
        return jsonObject.TryGetStringProperty(propertyName)
                         .Run()
                         .ThrowIfFail();
    }

    public static Eff<string> TryGetStringProperty(this JsonObject jsonObject, string propertyName)
    {
        return jsonObject.TryGetJsonValueOfTProperty<string>(propertyName);
    }

    public static Eff<T> TryGetJsonValueOfTProperty<T>(this JsonObject jsonObject, string propertyName)
    {
        return jsonObject.TryGetJsonValueProperty(propertyName)
                         .Bind(jsonValue => jsonValue.TryGetValue<T>(out var value)
                                            ? SuccessEff(value)
                                            : FailEff<T>(JsonModule.GetJsonError($"Property '{propertyName}' is not a JSON value of type '{typeof(T).Name}'.")));
    }

    public static Eff<JsonValue> TryGetJsonValueProperty(this JsonObject jsonObject, string propertyName)
    {
        return jsonObject.TryGetProperty(propertyName)
                         .Bind(node => node is JsonValue jsonValue
                                        ? SuccessEff(jsonValue)
                                        : FailEff<JsonValue>(JsonModule.GetJsonError($"Property '{propertyName}' is not a JSON value.")));
    }

    public static Eff<Seq<JsonObject>> TryGetJsonObjectArray(this JsonObject jsonObject, string propertyName)
    {
        return jsonObject.TryGetJsonArrayProperty(propertyName)
                         .Bind(jsonArray => jsonArray.Sequence(node => node is JsonObject jsonObject
                                                                        ? SuccessEff(jsonObject)
                                                                        : FailEff<JsonObject>(JsonModule.GetJsonError("Node is not a JSON object.")))
                                                     .MapFail(error => JsonModule.GetJsonError($"Property '{propertyName}' has at least one element that is not a JSON object.")))
                         .Map(enumerable => enumerable.ToSeq());
    }

    public static JsonArray GetJsonArrayProperty(this JsonObject jsonObject, string propertyName)
    {
        return jsonObject.TryGetJsonArrayProperty(propertyName)
                         .Run()
                         .ThrowIfFail();
    }

    public static Eff<JsonArray> TryGetJsonArrayProperty(this JsonObject jsonObject, string propertyName)
    {
        return jsonObject.TryGetProperty(propertyName)
                         .Bind(node => node is JsonArray jsonArray
                                        ? SuccessEff(jsonArray)
                                        : FailEff<JsonArray>(JsonModule.GetJsonError($"Property '{propertyName}' is not a JSON array.")));
    }

    public static JsonObject GetJsonObjectProperty(this JsonObject jsonObject, string propertyName)
    {
        return jsonObject.TryGetJsonObjectProperty(propertyName)
                         .Run()
                         .ThrowIfFail();
    }

    public static Eff<JsonObject> TryGetJsonObjectProperty(this JsonObject jsonObject, string propertyName)
    {
        return jsonObject.TryGetProperty(propertyName)
                         .Bind(node => node is JsonObject jsonObject
                                        ? SuccessEff(jsonObject)
                                        : FailEff<JsonObject>(JsonModule.GetJsonError($"Property '{propertyName}' is not a JSON object.")));
    }

    public static JsonNode GetProperty(this JsonObject jsonObject, string propertyName)
    {
        return jsonObject.TryGetProperty(propertyName)
                         .Run()
                         .ThrowIfFail();
    }

    public static Eff<JsonNode> TryGetProperty(this JsonObject jsonObject, string propertyName)
    {
        return jsonObject.TryGetOptionalProperty(propertyName)
                         .Bind(option => option.ToEff(JsonModule.GetJsonError($"Property '{propertyName}' does not exist in JSON object.")));
    }

    public static Option<JsonNode> GetOptionalProperty(this JsonObject jsonObject, string propertyName)
    {
        return jsonObject.TryGetOptionalProperty(propertyName)
                         .Run()
                         .ThrowIfFail();
    }

    public static Eff<Option<JsonNode>> TryGetOptionalProperty(this JsonObject jsonObject, string propertyName)
    {
        return jsonObject.GetOptionalNullableProperty(propertyName)
                         .Map(option => option.ToEff(JsonModule.GetJsonError($"Property '{propertyName}' has a null value.")))
                         .Sequence();
    }

    public static Option<Option<JsonNode>> GetOptionalNullableProperty(this JsonObject jsonObject, string propertyName)
    {
        return jsonObject.TryGetPropertyValue(propertyName, out var property)
            ? Some(property is null ? None : Some(property))
            : None;
    }

    public static Aff<JsonObject> FromStream(Stream stream, CancellationToken cancellationToken)
    {
        var serializerOptions = new JsonSerializerOptions { PropertyNameCaseInsensitive = true };

        return Aff(() => JsonSerializer.DeserializeAsync<JsonObject>(stream, serializerOptions, cancellationToken))
                .Bind(jsonObject => jsonObject is null
                                    ? FailAff<JsonObject>(JsonModule.GetJsonError("Stream has a null JSON object."))
                                    : SuccessAff(jsonObject));
    }

    public static Eff<JsonObject> FromBinaryData(BinaryData binaryData)
    {
        var serializerOptions = new JsonSerializerOptions(JsonSerializerDefaults.Web);

        return Eff(() => JsonSerializer.Deserialize<JsonObject>(binaryData, serializerOptions))
                .Bind(jsonObject => jsonObject is null
                                    ? FailEff<JsonObject>(JsonModule.GetJsonError("JSON object is null."))
                                    : SuccessEff(jsonObject));
    }

    public static string SerializeToString(this JsonObject jsonObject)
    {
        var serializerOptions = new JsonSerializerOptions(JsonSerializerDefaults.Web)
        {
            Encoder = JavaScriptEncoder.UnsafeRelaxedJsonEscaping
        };

        return jsonObject.ToJsonString(serializerOptions);
    }
}

File: code/common/Json.fs
namespace common

open FSharpPlus
open System
open System.IO
open System.Text.Encodings.Web
open System.Text.Json
open System.Text.Json.Nodes

[<RequireQualifiedAccess>]
type JsonError =
    private
    | JsonError of NonEmptyString

    static member fromString value =
        if String.IsNullOrWhiteSpace(value) then
            invalidOp "JSON error cannot be empty."
        else
            value |> NonEmptyString.fromString |> JsonError

    static member toString(JsonError (NonEmptyString value)) = value

[<RequireQualifiedAccess>]
module JsonObject =
    let private serializerOptions =
        let options = new JsonSerializerOptions()
        options.PropertyNameCaseInsensitive <- false
        options.Encoder <- JavaScriptEncoder.UnsafeRelaxedJsonEscaping
        options

    let tryFromStream (stream: Stream) =
        async {
            try
                let! cancellationToken = Async.CancellationToken

                let! jsonObject =
                    JsonSerializer.DeserializeAsync<JsonObject>(stream, serializerOptions, cancellationToken)
                    |> Async.fromValueTaskOf

                return
                    if isNull jsonObject then
                        "Stream has a null JSON object"
                        |> JsonError.fromString
                        |> Error
                    else
                        Ok jsonObject
            with
            | :? JsonException as jsonException ->
                return
                    jsonException.Message
                    |> JsonError.fromString
                    |> Error
            | error -> return raise error
        }

    let tryFromBytes (bytes: byte array) =
        try
            let jsonObject =
                JsonSerializer.Deserialize<JsonObject>(bytes, serializerOptions)

            if isNull jsonObject then
                "Stream has a null JSON object"
                |> JsonError.fromString
                |> Error
            else
                Ok jsonObject
        with
        | :? JsonException as jsonException ->
            jsonException.Message
            |> JsonError.fromString
            |> Error
        | error -> raise error

    let toBytes (jsonObject: JsonObject) =
        JsonSerializer.SerializeToUtf8Bytes(jsonObject, serializerOptions)

    let toString (jsonObject: JsonObject) =
        jsonObject.ToJsonString(serializerOptions)

    let getOptionalNullableProperty propertyName (jsonObject: JsonObject) =
        match jsonObject.TryGetPropertyValue(propertyName) with
        | true, value ->
            if isNull value then
                Some None
            else
                Some(Some value)
        | _ -> None

    let tryGetNullableProperty propertyName jsonObject =
        jsonObject
        |> getOptionalNullableProperty propertyName
        |> Option.toResultWith (JsonError.fromString $"Property '{propertyName}' does not exist in JSON object.")

    let tryGetProperty propertyName jsonObject =
        jsonObject
        |> tryGetNullableProperty propertyName
        |> Result.bind (Option.toResultWith (JsonError.fromString $"Property '{propertyName}' has a null value."))

    let tryGetJsonObjectProperty propertyName jsonObject =
        jsonObject
        |> tryGetProperty propertyName
        |> Result.bind (function
            | :? JsonObject as jsonObject -> Ok jsonObject
            | _ ->
                $"Property '{propertyName}' is not a JSON object."
                |> JsonError.fromString
                |> Error)

    let tryGetJsonArrayProperty propertyName jsonObject =
        jsonObject
        |> tryGetProperty propertyName
        |> Result.bind (function
            | :? JsonArray as jsonArray -> Ok jsonArray
            | _ ->
                $"Property '{propertyName}' is not a JSON array."
                |> JsonError.fromString
                |> Error)

    let tryGetJsonObjectArrayProperty propertyName jsonObject =
        jsonObject
        |> tryGetJsonArrayProperty propertyName
        |> Result.bind (
            Result.traverseSeq (function
                | null -> "Node is null." |> JsonError.fromString |> Error
                | :? JsonObject as jsonObject -> Ok jsonObject
                | _ ->
                    "Node is not a JSON object."
                    |> JsonError.fromString
                    |> Error)
        )
        |> Result.mapError (fun _ ->
            JsonError.fromString
                $"Property '{propertyName}' has an element that is either null or not a valid JSON object.")

    let tryGetJsonValueProperty propertyName jsonObject =
        jsonObject
        |> tryGetProperty propertyName
        |> Result.bind (function
            | :? JsonValue as jsonValue -> Ok jsonValue
            | _ ->
                $"Property '{propertyName}' is not a JSON value."
                |> JsonError.fromString
                |> Error)

    let tryGetJsonValuePropertyOf<'a> propertyName jsonObject =
        jsonObject
        |> tryGetJsonValueProperty propertyName
        |> Result.bind (fun jsonValue ->
            match jsonValue.TryGetValue<'a>() with
            | true, value -> Ok value
            | _ ->
                $"Property '{propertyName}' is not a JSON value of type '{typeof<'a>.Name}'."
                |> JsonError.fromString
                |> Error)

    let tryGetStringProperty propertyName jsonObject =
        jsonObject
        |> tryGetJsonValuePropertyOf<string> propertyName

    let tryGetNonEmptyStringProperty propertyName jsonObject =
        jsonObject
        |> tryGetStringProperty propertyName
        |> Result.bind (fun value ->
            if String.IsNullOrWhiteSpace(value) then
                $"Property '{propertyName}' is an empty or whtespace string."
                |> JsonError.fromString
                |> Error
            else
                value |> NonEmptyString.fromString |> Ok)

    let tryGetUIntProperty propertyName jsonObject =
        jsonObject
        |> tryGetJsonValuePropertyOf<uint> propertyName

    let addProperty propertyName node (jsonObject: JsonObject) =
        jsonObject.Add(propertyName, node)
        jsonObject

    let addStringProperty propertyName (value: string) jsonObject =
        jsonObject
        |> addProperty propertyName (JsonNode.op_Implicit (value))

[<RequireQualifiedAccess>]
module JsonArray =
    let addProperty node (jsonArray: JsonArray) =
        jsonArray.Add(node)
        jsonArray

    let addStringProperty (value: string) jsonArray =
        jsonArray
        |> addProperty (JsonNode.op_Implicit (value))
